GEP-0043: Spegel Registry Support in the registry-cache extension (#44)
gardener/enhancements: geps/0040-spegel-registry-support-in-the-registry-cache-extension/README.md

> # GEP-0040: Spegel Registry Support in the registry-cache extension

Reviewer: ```suggestion
# GEP-0043: Spegel Registry Support in the registry-cache extension
```

- [Appendix (Optional)](#appendix-optional)

## Motivation

The current solution for image cache is based on [Distribution project](https://github.com/distribution/distribution)'s pull-through cache.

### The Problem

The registry cache extension is not widely adopted in some Gardener landscapes, and not all used upstreams are covered by the extension's configurations. This is likely due to:
- **Explicit and non-trivial configuration**: A separate [configuration](https://github.com/gardener/gardener-extension-registry-cache/blob/v0.19.0/docs/usage/registry-cache/configuration.md#shoot-configuration) needs to be set for each upstream registry.
- **Additional resources**: A StatefulSet is created for each of the configured upstream. The PV size (defaults to 10Gi) should be monitored and [increased](https://github.com/gardener/gardener-extension-registry-cache/blob/v0.19.0/docs/usage/registry-cache/configuration.md#increase-the-cache-disk-size) if necessary, otherwise the cache becomes ineffective.
- **[Limitations](https://github.com/gardener/gardener-extension-registry-cache/blob/v0.19.0/docs/usage/registry-cache/configuration.md#limitations)**:

>   - It is not possible to cache most of images from the pods in the `kube-system` namespace, as the cache can be used only after its Kubernetes service can be accessed from the host (i.e., after kube-proxy configures iptables/IPVS rules).

Reviewer 2: Nit: You can have clusters without `kube-proxy` when service routing is implemented in CNI.

```suggestion
  - It is not possible to cache most of images from the pods in the `kube-system` namespace, as the cache can be used only after its Kubernetes service can be accessed from the host (i.e., after service routing was implemented, e.g. by kube-proxy configuring iptables/IPVS rules).
```

  - Only one set of [upstream credentials](https://github.com/gardener/gardener-extension-registry-cache/blob/v0.19.0/docs/usage/registry-cache/upstream-credentials.md#how-to-provide-credentials-for-upstream-registry) is supported, therefore some private images cannot be cached.
  - Amazon ECR mirroring is not supported - [ref](https://github.com/gardener/gardener-extension-registry-cache/issues/259).

>   - Dynatrace registry mirroring is not effective, as incorrect `Content-Length: 0` header is returned for most of the image layers.

Reviewer: Is this something that can only be fixed with Spegel?

  - [TTL](https://github.com/gardener/gardener-extension-registry-cache/blob/v0.19.0/docs/usage/registry-cache/configuration.md#garbage-collection) garbage collection setting is inefficient because it is not extended on access. Hence, an image is always deleted from the cache when its TTL expires, and will be re-pulled from upstream if requested again.

### Why Should We Care

Due to the low adoption rate and incomplete coverage of upstream registers, there are significant NAT Gateway traffic costs for downloading images.
Additionally, there are costs for outbound traffic from image registries like `Amazon Elastic Container Registry`, `Google Cloud Artifact Registry`, and `Azure Container Registry`.
Even if the pull-through cache is configured, there are also costs for cross zonal traffic within the Shoot cluster.

> Providing a p2p image caching solution that is enabled by default and spans all upstream registries will reduce both cloud costs and image pull latency.

Reviewer 2: Nit:

```suggestion
Providing a peer-to-peer (p2p) image caching solution that is enabled by default and spans all upstream registries will reduce both cloud costs and image pull latency.
```

### Who Benefits

- **End users** benefit from the reduced latency when pulling images.
- **Stakeholders** see reduced costs for NAT Gateways and registries.
- **Gardener operators** are no longer engaged in monitoring notifications of PV sizes in the registry caches.

## Proposal

> The proposal is to extend the existing [registry-cache](https://github.com/gardener/gardener-extension-registry-cache) extension with a p2p image cache based on Spegel. A new extension type `registry-spegel` will be enabled by default for all Shoot clusters and will be used as the [`_default`][setup-default-mirror-for-all-registries] mirror configuration in `containerd`.

Reviewer 2: Do you know of any real scenarios where you would prefer the old `registry-cache` over one built with spegel? I can make up scenarios where all nodes go away and images are huge so that the persistent volume would be helpful, but that does not seem to be practical at all.
In other words, do you really believe that the old `registry-cache` is required in the long run if we have support for spegel?

### Introduction

A brief overview of the components used in this proposal.

#### Spegel overview

[Spegel][spegel] is a p2p image cache. It runs a local image registry on each node in a Kubernetes cluster. The local registry subscribes for containerd's image related events and serves image blobs and manifests that are available in the containerd image store.
The content discovery in a Kubernetes cluster is based on [Kademlia DHT][content-provider-routing]. When an image content (blob, manifest or index) is available in the containerd image store, Spegel adds its `digest` to the DHT provider store, announcing that the node provides the content corresponding to the `digest`. Then, when the same `digest` is needed by another node, Spegel searches the DHT for peers that provide the content and pulls it from them.

> The straightforward way to deploy Spegel to a Kubernetes cluster is by using the provided helm chart. However, this has some [drawbacks](https://spegel.dev/docs/faq/#what-should-i-do-if-other-pods-are-scheduled-on-new-nodes-before-spegel) when a new node join the cluster. 

Reviewer 2: ```suggestion
The straightforward way to deploy Spegel to a Kubernetes cluster is by using the provided helm chart. However, this has some [drawbacks](https://spegel.dev/docs/faq/#what-should-i-do-if-other-pods-are-scheduled-on-new-nodes-before-spegel) when a new node joins the cluster. 
```

> Our goal is to be able to use Spegel for all images pulled from the kubelet, including the `registry.k8s.io/pause` image. Therefore it was decided to [run][run-spegel-on-host] Spegel registry as a systemd unit service on the node. This requires contributing a new Spegel `bootstrapper` or extending the existing [HTTP bootstrapper](https://github.com/spegel-org/spegel/blob/6f02215fa3fc1d3bbdb11fa62dfa7c07dbe3b7c2/pkg/routing/bootstrap.go#L131-L135).

Reviewer 2: Just to be fully transparent, what container images do we expect to be pulled without spegel? I suppose `gardener-node-agent` will not be pulled through spegel, but I am not sure about other container images, e.g. `kubelet`.
Could you please share the set of container images, which will definitively not be pulled through spegel?

	// `routerPort` should be a valid port number (1-65535, inclusive).
	// Defaults to 5001.
	// +optional
	RouterPort *int32 `json:"routerPort,omitempty"`
	// MetricsPort is the metrics port on each Node.
	// `metricsPort` should be a valid port number (1-65535, inclusive).
	// Defaults to 9090.
	// +optional
	MetricsPort *int32 `json:"metricsPort,omitempty"`
}
```

### Architecture Overview 

The `registry-spegel` extension type will be enabled by default for all Shoot clusters in the operator extension resource:
```yaml
apiVersion: operator.gardener.cloud/v1alpha1
kind: Extension
metadata:
  name: extension-registry-cache
  annotations:
    security.gardener.cloud/pod-security-enforce: baseline
spec:
  resources:
  - kind: Extension
    type: registry-spegel
    autoEnable: [shoot]
    clusterCompatibility: [shoot]
  - kind: Extension
    type: registry-cache
    clusterCompatibility: [shoot]
  - kind: Extension
    type: registry-mirror
    clusterCompatibility: [shoot]
...
```
Spegel `bootstrapper` is deployed into the Shoot namespace in the Seed cluster. In the Shoot cluster nodes the `containerd`, `spegel`, and `kubelet` systemd units are started.

> When `spegel` is starting, it gets the bootstrap peers from the `bootstrapper` and registers for image related events in `containerd`. It also lists existing content in the image store and advertises it to the DHT.

Reviewer: What happens when `kubelet` already starts to pull an image while Spegel has not yet contact (or is still in the process of contacting) the `bootstrapper`?
